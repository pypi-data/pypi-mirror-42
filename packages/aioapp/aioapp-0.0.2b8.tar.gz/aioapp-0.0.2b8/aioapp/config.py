import copy
import os
from collections import OrderedDict
from typing import Optional, Union, Dict, Any, Type
from os import _Environ

Env = Union[_Environ, dict]


class ConfigError(Exception):
    pass


class Val:

    def __init__(self, name: str, value: Any) -> None:
        self.name = name
        self.value = value

    def __call__(self):  # pragma: nocover
        raise NotImplementedError()

    @staticmethod
    def type_name() -> str:
        return ''

    def args_markdown(self) -> str:
        return ''


class StrVal(Val):

    def __init__(self, name: str, value: Any,
                 max: Optional[int] = None,
                 min: Optional[int] = None) -> None:
        super().__init__(name, value)
        self.min = min
        self.max = max

    def __call__(self) -> str:
        if not isinstance(self.value, str):
            raise ConfigError("%s must be a string" % self.name)
        if self.min is not None and len(self.value) < self.min:
            raise ConfigError("length of %s must be greater than or equal to "
                              "%s"
                              "" % (self.name, self.min))
        if self.max is not None and len(self.value) > self.max:
            raise ConfigError("length of %s must be less than or equal to %s"
                              "" % (self.name, self.max))

        return self.value

    @staticmethod
    def type_name() -> str:
        return 'string'

    def args_markdown(self) -> str:
        text = ''
        if self.min is not None:
            text += '\n  min length: %s' % self.min
        if self.max is not None:
            text += '\n  max length: %s' % self.max
        return text


class BoolVal(Val):

    def __call__(self) -> bool:
        if isinstance(self.value, bool):
            return self.value
        if isinstance(self.value, int):
            return self.value != 0
        if isinstance(self.value, str):
            if self.value.lower() in ('1', 'on', 'true', 't'):
                return True
            if self.value.lower() in ('0', 'off', 'false', 'f'):
                return False
        raise ConfigError("%s must be a boolean" % self.name)

    @staticmethod
    def type_name() -> str:
        return 'boolean'


class IntVal(Val):

    def __init__(self, name: str, value: Any,
                 max: Optional[int] = None,
                 min: Optional[int] = None) -> None:
        super().__init__(name, value)
        self.min = min
        self.max = max

    def __call__(self) -> int:
        try:
            val = int(self.value)
        except Exception:
            raise ConfigError("%s must be an integer" % self.name)
        if self.min is not None and val < self.min:
            raise ConfigError("%s must be greater than or equal to %s"
                              "" % (self.name, self.min))
        if self.max is not None and val > self.max:
            raise ConfigError("%s must be less than or equal to %s"
                              "" % (self.name, self.max))
        return val

    @staticmethod
    def type_name() -> str:
        return 'integer'

    def args_markdown(self) -> str:
        text = ''
        if self.min is not None:
            text += '\n  min value: %s' % self.min
        if self.max is not None:
            text += '\n  max value: %s' % self.max
        return text


class FloatVal(Val):

    def __init__(self, name: str, value: Any,
                 max: Optional[float] = None,
                 min: Optional[float] = None) -> None:
        super().__init__(name, value)
        self.min = min
        self.max = max

    def __call__(self) -> float:
        try:
            val = float(self.value)
        except Exception:
            raise ConfigError("%s must be a float" % self.name)
        if self.min is not None and val < self.min:
            raise ConfigError("%s must be greater than or equal to %s"
                              "" % (self.name, self.min))
        if self.max is not None and val > self.max:
            raise ConfigError("%s must be less than or equal to %s"
                              "" % (self.name, self.max))
        return val

    @staticmethod
    def type_name() -> str:
        return 'float'

    def args_markdown(self) -> str:
        text = ''
        if self.min is not None:
            text += '\n  min value: %s' % self.min
        if self.max is not None:
            text += '\n  max value: %s' % self.max
        return text


class FileVal(Val):

    def __init__(self, name: str, value: Any,
                 mode: str = 'r',
                 encoding: str = 'UTF-8') -> None:
        super().__init__(name, value)
        self.mode = mode
        self.encoding = encoding

    def __call__(self, mode: str = 'r', encoding: str = 'UTF-8') -> str:
        try:
            with open(self.value, mode, encoding=encoding):
                pass
        except Exception as e:
            raise ConfigError("Could not access to file %s with error: %s"
                              "" % (self.value, e))
        return self.value

    @staticmethod
    def type_name() -> str:
        return 'string(path to file)'

    def args_markdown(self) -> str:
        text = ''
        if self.mode is not None:
            text += '\n  assess mode: %s' % self.mode
        if self.encoding is not None:
            text += '\n  encoding: %s' % self.encoding
        return text


class DirVal(Val):

    def __call__(self) -> str:
        if not os.path.exists(self.value):
            raise ConfigError("Directory %s does not exist"
                              "" % self.value)
        return self.value

    @staticmethod
    def type_name() -> str:
        return 'string(path to dir)'


class Config:
    _vars: Dict[str, Union[Dict, OrderedDict]] = {}

    def __init__(self,
                 env: Optional[Env] = None) -> None:
        self._env = env or os.environ
        self._conf = copy.deepcopy(self._vars)
        self._description: Dict[str, Dict] = {}
        for key, val in self._conf.items():
            val_type = val.pop('type')
            val_name = val.pop('name')
            val_default = None
            if 'default' in val:
                val_default = val.pop('default')
            required = False
            if 'required' in val:
                required = bool(val.pop('required'))
            descr = ''
            if 'descr' in val:
                descr = str(val.pop('descr'))

            self._description[val_name] = {
                'type': val_type,
                'default': val_default,
                'required': required,
                'descr': descr,
            }

            value = self._env.get(val_name, val_default)
            if value is None:
                if required is True:
                    raise ConfigError("%s is required" % val_name)
                else:
                    setattr(self, key, None)
            else:
                v: Any = self._get_val(val_type)
                setattr(self, key, v(val_name, value, **val)())

    @classmethod
    def _get_val(cls, val_type: Any) -> Type[Val]:
        if val_type == str:
            return StrVal
        elif val_type == bool:
            return BoolVal
        elif val_type == int:
            return IntVal
        elif val_type == float:
            return FloatVal
        elif val_type == 'file':
            return FileVal
        elif val_type == 'dir':
            return DirVal
        elif isinstance(val_type, type) and issubclass(val_type, Val):
            return val_type
        else:
            raise UserWarning('Invalid configuration settings')

    @classmethod
    def as_markdown(cls):
        result = []

        for name, options in cls._vars.items():
            options = copy.deepcopy(options)

            type_cls = cls._get_val(options.pop('type'))
            env_name = options.pop('name')
            descr = ''
            if 'descr' in options:
                descr = ': %s' % options.pop('descr')
            text = '* %s%s\n  type: %s' % (env_name, descr,
                                           type_cls.type_name())
            if 'required' in options:
                required = options.pop('required')
                if required:
                    text += '\n\n  required'
            if 'default' in options:
                default = options.pop('default')
                if default:
                    text += '\n  default: %s' % default

            inst = type_cls(env_name, None, **options)
            text += inst.args_markdown()

            result.append(text)

        return '\n\n'.join(result) + '\n'
