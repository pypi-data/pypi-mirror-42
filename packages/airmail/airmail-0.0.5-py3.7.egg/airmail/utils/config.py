from functools import reduce

def build_config(transforms, config):
    """Reduces a dictionary through functions to modify the dictionary"""
    return reduce(lambda c, fn: fn(c), transforms, config)
