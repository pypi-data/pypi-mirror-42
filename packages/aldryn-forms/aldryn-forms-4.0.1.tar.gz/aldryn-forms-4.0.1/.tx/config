[main]
host = https://www.transifex.com

[aldryn-forms.aldryn_forms]
file_filter = aldryn_forms/locale/<lang>/LC_MESSAGES/django.po
source_file = aldryn_forms/locale/en/LC_MESSAGES/django.po
source_lang = en
type = PO

