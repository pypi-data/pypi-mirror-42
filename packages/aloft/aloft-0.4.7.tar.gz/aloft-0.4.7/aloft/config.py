from aloft import input


def get_profile():
    return get_aloft_config()['profile']


def get_roles():
    return get_aloft_config()['roles']


def get_aloft_config():
    return input.read_yaml_file(get_aloft_config_filename())


def get_aloft_config_filename():
    aloft_config_directory = get_aloft_config_directory()
    return f'{aloft_config_directory}/config.yaml'


def get_aloft_config_directory():
    return input.get_directory_from_environment('ALOFT_CONFIG', '~/.aloft')
