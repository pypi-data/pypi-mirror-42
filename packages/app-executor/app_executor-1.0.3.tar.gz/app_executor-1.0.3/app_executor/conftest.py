import pytest
import app_executor


@pytest.yield_fixture()
def executor(tmpdir):
    """
    return app_executor.AppExecutor object that will allow to run arbitrary number of
    processes. In teardown it will stop them and perform dump analysis. All results will be
    put into temporary directory.
    """
    with app_executor.AppExecutor(tmpdir) as app_exec:
        yield app_exec
