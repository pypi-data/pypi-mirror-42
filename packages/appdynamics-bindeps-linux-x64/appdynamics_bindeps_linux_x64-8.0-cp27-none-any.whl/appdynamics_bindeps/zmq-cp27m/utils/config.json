{
  "skip_check_zmq": false, 
  "no_libzmq_extension": true, 
  "build_ext": {}, 
  "have_sys_un_h": true, 
  "zmq_prefix": "/python-agent/bindeps/deps/libzmq/binaries/release", 
  "libzmq_extension": true, 
  "bdist_egg": {}
}