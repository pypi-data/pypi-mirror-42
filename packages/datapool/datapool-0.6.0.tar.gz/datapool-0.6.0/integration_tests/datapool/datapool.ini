[db]
connection_string = sqlite+pysqlite:////Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz/.simple.db

[backup_landing_zone]
folder = 

[r]
extension = .r
executable = /usr/local/bin/R

[matlab]
extension = .m
executable = /Applications/MATLAB_R2016a.app/bin/matlab

[julia]
extension = .jl
version = 0.5.2
executable = /usr/local/bin/julia

[python]
extension = .py
executable = /Users/uweschmitt/Projects/eawag_datapool/venv/bin/python3

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[landing_zone]
folder = /Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz

