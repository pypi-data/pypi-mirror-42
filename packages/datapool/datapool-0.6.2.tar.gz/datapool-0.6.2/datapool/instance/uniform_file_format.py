# encoding: utf-8
from __future__ import print_function, division, absolute_import

import csv
import datetime
import io
import os

from datapool.utils import iter_to_list


def _check_header(header):

    header = list(header)

    if "site" in header:
        header.remove("site")
    else:
        for n in "xyz":
            if n not in header:
                raise ValueError(
                    "header {!r} must have either 'site' or 'x', 'y' and 'z'".format(
                        header
                    )
                )
            header.remove(n)

    header = set(header)
    fix_column_names = set(("timestamp", "parameter", "value"))
    if header != fix_column_names:
        missing = ["'{}'".format(name) for name in fix_column_names - header]
        invalid = ["'{}'".format(name) for name in header - fix_column_names]
        msg = ""
        if missing:
            msg += "missing: {}".format(", ".join(sorted(missing)))
        if invalid:
            msg += "invalid: {}".format(", ".join(sorted(invalid)))
        raise ValueError("invalid header ({})".format(msg))


@iter_to_list
def read_from_file(path):
    if not os.path.exists(path):
        raise ValueError("{} does not exist".format(path))

    with open(path, "r", encoding="ascii") as fh:
        yield from _read_from_fh(fh)


@iter_to_list
def read_from_string(data):
    fh = io.StringIO(data)
    yield from _read_from_fh(fh)


def can_be_converted_to_float(value):
    try:
        value = float(value)
    except ValueError:
        return False
    return True


def parse_timestamp(timestamp):
    return datetime.datetime.strptime(timestamp, "%Y-%m-%d %H:%M:%S")


def _read_from_fh(fh):
    reader = csv.reader(fh, delimiter=";")
    header = [f.strip() for f in next(reader)]
    _check_header(header)
    for i, row in enumerate(reader):
        if len(row) < len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is incomplete".format(i + 1, row))
        if len(row) > len(header):
            # i + 1, we skipped header
            raise ValueError("row {} ({!r}) is to long".format(i + 1, row))

        row = (cell.strip() for cell in row)
        row_dict = dict(zip(header, row))
        yield row_dict


@iter_to_list
def check_rows(row_dicts):

    if not row_dicts:
        return

    for i, row_dict in enumerate(row_dicts):

        value = row_dict["value"]
        if not can_be_converted_to_float(value):
            yield "row {}: value '{}' is not a float".format(i, value)
        else:
            row_dict["value"] = float(value)

        timestamp = row_dict["timestamp"]
        try:
            row_dict["timestamp"] = parse_timestamp(timestamp)
        except ValueError:
            yield "row {}: timstamp '{}' either has wrong format or is invalid".format(
                i, timestamp
            )

        has_xyz = all(name in row_dict and row_dict[name] is not None for name in "xyz")
        not_empty = ("parameter",)
        if not has_xyz:
            not_empty += ("site",)

        for name in not_empty:
            if row_dict[name].strip() == "":
                yield "row {}: field {} is empty".format(i, name)


@iter_to_list
def to_signals(row_dicts):

    for row in row_dicts:
        for f in "xyz":
            if f in row:
                row["coord_" + f] = row[f]
                del row[f]

    from .domain_objects import Signal

    return list(map(Signal, row_dicts))
