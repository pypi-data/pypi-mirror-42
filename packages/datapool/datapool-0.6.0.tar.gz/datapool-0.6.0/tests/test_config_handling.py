# encoding: utf-8
from __future__ import absolute_import, division, print_function

import os

from datapool.config import read_ini, write_ini
from datapool.instance.config_handling import (
    check_config,
    check_folder,
    config_ini_file_path,
    init_config,
)

from .conftest import not_for_root


def _check_and_cleanup_config_for_regtest(config):

    assert len(config.db.connection_string) > 0
    config.db.connection_string = "<dummy for regression tests>"

    if "landing_zone" in config:
        assert config.landing_zone.folder != ""
        config.landing_zone.folder = "<dummy for regression tests>"
    if "backup_landing_zone" in config:
        if config.backup_landing_zone.folder != "":
            config.backup_landing_zone.folder = "<dummy for regression tests>"

    config.r.executable = " <dummy for regression tests>"
    config.matlab.executable = " <dummy for regression tests>"
    config.julia.executable = " <dummy for regression tests>"


def test_faked_config(regtest, here):
    path = config_ini_file_path()

    # relpath to avoid absolute pathes in regtest output:
    print(os.path.relpath(path, here), file=regtest)
    config = read_ini(path)

    _check_and_cleanup_config_for_regtest(config)

    write_ini(config, regtest)


def test_init_config(
    monkeypatch, tmpdir, regtest, data_path, has_matlab, has_julia, has_r
):

    # we patch, else we could overwrite something, eg the files in
    # ./home/datapool which is our faked home folder for testing:
    monkeypatch.setenv("ETC", tmpdir.strpath)

    folder, messages = init_config(data_path("test_landing_zone"))
    assert folder is not None
    print(sorted(os.listdir(tmpdir.strpath)), file=regtest)

    config = read_ini(config_ini_file_path())

    if has_matlab:
        assert not any("matlab" in message for message in messages)
        assert config.matlab.executable != "", str(config)
    else:
        assert any("matlab" in message for message in messages)
        assert config.matlab.executable == "", str(config)

    if has_julia:
        assert not any("julia" in message for message in messages)
        assert config.julia.executable != "", str(config)
    else:
        assert any("julia" in message for message in messages)
        assert config.julia.executable == "", str(config)

    if has_r:
        assert not any("r" in message for message in messages)
        assert config.r.executable != "", str(config)
    else:
        assert any("r" in message for message in messages)
        assert config.r.executable == "", str(config)

    assert config.python.executable != ""
    del config.python.executable

    _check_and_cleanup_config_for_regtest(config)

    write_ini(config, regtest)


def test_check_config(config, regtest, tmpdir, regtest_logger):

    os.makedirs(config.landing_zone.folder)
    with regtest_logger():
        messages = check_config(config)

    for message in messages:
        print(message, file=regtest)

    # we now corrupt the config
    config.worker.port = 22  # should be in use
    config.log_receiver.port = 22.23  # should be int
    config.db.connection_string = ""  # must be set
    config.r.extension = ""  # must be set
    config.r.executable = 0  # wrong type
    config.matlab.extension = ""  # must be set
    config.matlab.executable = 0  # wrong type
    config.julia.extension = ""  # must be set
    config.julia.executable = 0  # wrong type
    config.logging.config_file = "./non_existing.yaml"

    with regtest_logger():
        messages = check_config(config)

    for message in messages:
        print(message, file=regtest)


@not_for_root
def test_check_folder(tmpdir, regtest):
    def check(folder):
        folder = folder.strpath
        result = check_folder(folder)
        if result is None:
            result = "ok"
        print("check {}: {}".format(folder, result), file=regtest)

    fldr = tmpdir.join("fresh_empty_writable_folder").mkdir()
    check(fldr)

    # not readable:
    fldr.chmod(0)
    check(fldr)

    # +rx, so not writable:
    fldr.chmod(0o555)
    check(fldr)

    # +wx, folder not empty
    fldr.chmod(0o777)
    fldr.join("file").open("w").close()
    check(fldr)
