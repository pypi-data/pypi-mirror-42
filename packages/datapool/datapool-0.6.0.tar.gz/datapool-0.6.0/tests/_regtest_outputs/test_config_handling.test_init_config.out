['datapool']
[db]
connection_string = <dummy for regression tests>

[backup_landing_zone]
folder = 

[r]
extension = .r
executable =  <dummy for regression tests>

[matlab]
extension = .m
executable =  <dummy for regression tests>

[julia]
extension = .jl
version = 1.0.1
executable =  <dummy for regression tests>

[python]
extension = .py

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[http_server]
port = 8000

[conversion]
block_size = 20000

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[datapool]
__version__ = 0.6.0

[landing_zone]
folder = <dummy for regression tests>

